### PR TITLE
Handle recursive types in the `data.jsondata` compiler plugin

### LINKED ISSUE #7008: Receive compilation error while using data.jsondaata moduel
**Description:**
When building the ballerina code[1] using `bal build` with the latest jsondata timestamp receive this compiler error.
```
error: compilation failed: The compiler extension in package 'ballerina:data.jsondata:0.3.0' failed to complete. null
```
**Steps to reproduce:**
[1] Ballerina code change done for http module:  https://github.com/ballerina-platform/module-ballerina-http/pull/2151

Priority/Blocker 

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #7010: `Stackoverflow` in the compiler plugin of `data.jsondata` module
Refer to https://github.com/ballerina-platform/ballerina-library/issues/7008

### PR DESCRIPTION
**Pull request by SasinduDilshara**
## Purpose

Fixes: https://github.com/ballerina-platform/ballerina-library/issues/7008
Fixes: https://github.com/ballerina-platform/ballerina-library/issues/7010

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/lib/data/jsondata/compiler/JsondataTypeValidator.java [hunk 1]
```diff
@@ -155,22 +156,31 @@ private void processFunctionDefinitionNode(FunctionDefinitionNode functionDefini
     }
 
     private void checkTypeAndDetectDuplicateFields(TypeSymbol typeSymbol, SyntaxNodeAnalysisContext ctx) {
+        checkTypeAndDetectDuplicateFields(typeSymbol, ctx, new HashSet<>());
+    }
+
+    private void checkTypeAndDetectDuplicateFields(TypeSymbol typeSymbol,
```
Comment (lnash94) [inline compiler-plugin/src/main/java/io/ballerina/lib/data/jsondata/compiler/JsondataTypeValidator.java:162]: Shall we have tests for these scenarios?

**Review by TharmiganK — COMMENTED** [review]
LGTM
If possible, shall we add tests as well?

File: compiler-plugin/src/main/java/io/ballerina/lib/data/jsondata/compiler/JsondataTypeValidator.java (hunk 1 above)
Comment (SasinduDilshara) [reply compiler-plugin/src/main/java/io/ballerina/lib/data/jsondata/compiler/JsondataTypeValidator.java:162]: There are already tests regarding this. I added another one.

**Comment by SasinduDilshara** [comment]
> LGTM If possible, shall we add tests as well?

There are testcases for scenarios.
There is a need to add tests for union type symbols.
Will add that as a separate task

**Review by lnash94 — APPROVED** [review]

**Comment by SasinduDilshara** [comment]
> > LGTM If possible, shall we add tests as well?
> 
> There are testcases for scenarios. There is a need to add tests for union type symbols. Will add that as a separate task

Will track in https://github.com/ballerina-platform/ballerina-library/issues/7015

**Pull request merged**